sources/web/datalab/authentication.ts

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2015 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/// <reference path="../../../externs/ts/node/node.d.ts" />
+/// <reference path="../../../externs/ts/node/node-cache.d.ts" />
+/// <reference path="common.d.ts" />
+
+
+import http = require('http');
+import httpapi = require('./httpapi');
+import logging = require('./logging');
+import NodeCache = require("node-cache");
+
+/**
+ * The cache service for whitelisted user.
+ */
+var authCache: NodeCache = new NodeCache();
+
+/**
+ * The application settings instance.
+ */
+var appSettings: common.Settings;
+
+function getAccessTokenFromMetadataServer(cb: common.Callback<string>) {
+  var headers: common.Map<string> = {'X-Google-Metadata-Request': 'True'};
+  try {
+    var tokenHost: string = "metadata.google.internal";
+    if (appSettings.metadataHost) {
+      tokenHost = appSettings.metadataHost;
+    }
+    httpapi.get(tokenHost, '/computeMetadata/v1/instance/service-accounts/default/token',
+        null, null, headers, function(e: Error, data: any) {
+      if (e) {
+        logging.getLogger().error(e, 'Failed to get access token from metadata service.');
+        cb && cb(e, null);
+        return;
+      }
+      if (data && data.access_token) {
+        cb && cb(null, data.access_token);
+      }
+      else {
+        logging.getLogger().error('Failed to get access token metadata service response.');
+        cb && cb(Error('metadata server'), null);
+      }
+    });
+  }
+  catch (e) {
+    logging.getLogger().error(e, 'Failed to get access token from metadata service.');
+    cb && cb(e, null);
+  }
+}
+
+function lookupUserFromDatastore(userId: string, accessToken: string, cb: common.Callback<boolean>) {
+  if (authCache.get(userId)) {
+    cb(null, true);
```

Comment:
We should invoke this asynchronously, so the callback is always executed in the same way.

    process.nextTick(function() {
      cb(null, true);
    });